pkg/conversion/deferred_converter.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/conversion"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+// deferredConverter implements conversion.Converter and determines on the fly what type of converter to use.
+type deferredConverter struct {
+	crd          *apiextensionsv1.CustomResourceDefinition
+	nopConverter conversion.CRConverter
+
+	// lock guards delegate
+	lock     sync.Mutex
+	delegate conversion.CRConverter
+
+	getAPIConversion func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIConversion, error)
+	newConverter     func(crd *apiextensionsv1.CustomResourceDefinition, apiConversion *apisv1alpha1.APIConversion) (conversion.CRConverter, error)
+}
+
+// Convert converts in to targetGV. If there is an APIConversion for this CRD, a CEL-based converter is used. Otherwise,
+// a nop converter is used.
+func (f *deferredConverter) Convert(in *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
+	converter, err := f.getConverter()
+	if err != nil {
+		return nil, fmt.Errorf("error getting converter: %w", err)
+	}
+
+	return converter.Convert(in, targetGV)
+}
+
+func (f *deferredConverter) getConverter() (conversion.CRConverter, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if f.delegate != nil {
+		return f.delegate, nil
```

Comment:
this surprises. What is a delegate in practice?

Reply:
i.c., you cache the result. Maybe add a oneliner comment, or rename to  `cachedConverter` or similar.

Reply:
I originally did this because at the time, I was copying the APIConversion from the source ws into system:bound-crds, and there was no guarantee the APIConversion was in the informer's cache when the serving info is created for the bound CRD (the converter is created at that time). But now that I'm not copying, we could consider removing this deferred converter entirely. The only upside of keeping it is that if you create a normal CRD and later you create an APIConversion, this deferredConverter allows the APIConversion to take effect. Otherwise, you have to deleted the CRD and recreate it.